File: src/0414-Third-Maximum-Number/third_maximum_number.cpp
#include <iostream>
#include <vector>
#include <climits>

class Solution {
public:
    int thirdMax(std::vector<int>& nums) {
        long first_max = LONG_MIN;
        long second_max = LONG_MIN;
        long third_max = LONG_MIN;

        for (int num : nums) {
            if (num > first_max) {
                third_max = second_max;
                second_max = first_max;
                first_max = num;
            } else if (num < first_max && num > second_max) {
                third_max = second_max;
                second_max = num;
            } else if (num < second_max && num > third_max) {
                third_max = num;
            }
        }

        return (third_max != LONG_MIN) ? third_max : first_max;
    }
};


int main() {
    std::vector<int> nums = {2,2,3,1};
    
    Solution solution;

    std::cout << solution.thirdMax(nums);
    
    return 0;
}

File: src/0414-Third-Maximum-Number/third_maximum_number.py
from typing import List

class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        first_max: int
        second_max: int
        third_max: int

        first_max = second_max = third_max = float('-inf')

        for num in nums:
            if num > first_max:
                first_max, second_max, third_max = num, first_max, second_max
            elif first_max > num > second_max:
                second_max, third_max = num, second_max
            elif second_max > num > third_max:
                third_max = num

        return third_max if third_max != float('-inf') else first_max


if __name__ == "__main__":
    nums: List[int] = [2,2,3,1]
    solution = Solution()
    
    print(solution.thirdMax(nums))